applications/FluidDynamicsApplication/custom_processes/shock_detection_process.h

```diff
@@ -0,0 +1,366 @@
+        // This is required for the nodal gradients calculation
+        if (mUpdateNodalAreaAtEachStep) {
+            CalculateNodalAreaProcess<CalculateNodalAreaSettings::SaveAsNonHistoricalVariable>(
+            mrModelPart,
+            mrModelPart.GetProcessInfo().GetValue(DOMAIN_SIZE)).Execute();
+        }
+
+        // If required recompute the NODAL_NEIGHBOURS
+        if (mUpdateNodalNeighboursAtEachStep) {
+            const auto& r_data_communicator = mrModelPart.GetCommunicator().GetDataCommunicator();
+            FindGlobalNodalNeighboursProcess(r_data_communicator, mrModelPart).Execute();
+        }
+
+        // Calculate the shock variable nodal gradients
+        ComputeNodalGradientProcess<ComputeNodalGradientProcessSettings::SaveAsNonHistoricalVariable>(
+            mrModelPart,
+            rShockVariable,
+            rShockGradientVariable).Execute();
+
+        auto& r_comm = mrModelPart.GetCommunicator();
+        auto& r_data_comm = r_comm.GetDataCommunicator();
+
+        // Create the global pointers list
+        GlobalPointersVector<Node<3>> global_pointers_list;
+        if (r_comm.IsDistributed()) {
+            for (auto &r_node : r_comm.LocalMesh().Nodes()) {
+                auto& r_gp_to_neighbours = r_node.GetValue(NEIGHBOUR_NODES).GetContainer();
+                for (auto &r_gp : r_gp_to_neighbours) {
+                    global_pointers_list.push_back(r_gp);
+                }
+            }
+            global_pointers_list.Unique();
+        }
+
+        // Now create the pointer communicator and shock values retrieve proxy
+        GlobalPointerCommunicator<Node<3>> pointer_communicator(r_data_comm, global_pointers_list);
+        auto shock_variables_proxy = pointer_communicator.Apply([&](const GlobalPointer<Node<3>>& rpNode)
+        {
+            NeighbourData<TShockVariableType, TShockGradientVariableType> neighbour_data(
+                rpNode->FastGetSolutionStepValue(rShockVariable),
+                rpNode->GetValue(rShockGradientVariable),
+                rpNode->Coordinates());
+            return neighbour_data;
+        });
+
+        // Perform the shock detection
+#pragma omp parallel for
+        for (int i_node = 0; i_node < static_cast<int>(r_comm.LocalMesh().NumberOfNodes()); ++i_node) {
+            auto it_node = r_comm.LocalMesh().NodesBegin() + i_node;
+            double& r_shock_sens = it_node->GetValue(mrShockSensorVariable);
+            const auto& r_var_i = it_node->FastGetSolutionStepValue(rShockVariable);
+            const auto& r_grad_var_i = it_node->GetValue(rShockGradientVariable);
+
+            // Loop the neighbours to compute the shock sensor
+            r_shock_sens = 0.0;
+            const double zero_tol = 1.0e-8;
+            auto& r_neighbours = it_node->GetValue(NEIGHBOUR_NODES);
+            KRATOS_DEBUG_ERROR_IF(r_neighbours.size() == 0) << "Node " << i_node << " has no neighbours." << std::endl;
+            for (auto& r_neigh : r_neighbours ) {
```

Comment:
are u sure that the node itself is not included in the neighbours?

Reply:
Sure. I checked it.